presto-main/src/test/java/io/prestosql/sql/query/QueryAssertions.java

```diff
@@ -280,14 +289,68 @@ private QueryAssert matches(MaterializedResult expected)
+                                    matchWithTolerance(actualRow, expectedRow, tolerances);
+                                });
+                    }
+                    else {
+                        assertion.containsExactlyElementsOf(expected.getMaterializedRows());
+                    }
                 }
                 else {
+                    if (tolerances != null && !tolerances.isEmpty()) {
+                        throw new UnsupportedOperationException("can not support in exact matches with unordered results.");
+                    }
                     assertion.containsExactlyInAnyOrderElementsOf(expected.getMaterializedRows());
                 }
             });
         }
 
+        private void matchWithTolerance(MaterializedRow actualRow, MaterializedRow expectedRow, List<Optional<Number>> tolerances)
+        {
+            assertEquals(actualRow.getFieldCount(), expectedRow.getFieldCount());
+
+            IntStream.range(0, actualRow.getFieldCount())
+                    .forEach(fieldIndex -> {
+                        Object actualRowField = actualRow.getField(fieldIndex);
+                        Object expectedRowField = expectedRow.getField(fieldIndex);
+
+                        if (tolerances.get(fieldIndex).isPresent()) {
+                            Number tolerance = tolerances.get(fieldIndex).get();
+                            assertTrue(expectedRowField instanceof Number && expectedRowField instanceof Comparable,
+                                    "tolerance is specified for the column but column is not a number and comparable " + expectedRowField);
+
+                            Range range = null;
+                            if (tolerance instanceof Integer) {
+                                range = Range.open(((Number) expectedRowField).intValue() - (Integer) tolerance, ((Number) expectedRowField).intValue() + (Integer) tolerance);
+                            }
+                            else if (tolerance instanceof Long) {
+                                range = Range.open(((Number) expectedRowField).longValue() - (Long) tolerance, ((Number) expectedRowField).longValue() + (Long) tolerance);
+                            }
+                            else if (tolerance instanceof Float) {
+                                range = Range.open(((Number) expectedRowField).floatValue() - (Float) tolerance, ((Number) expectedRowField).floatValue() + (Float) tolerance);
+                            }
+                            else if (tolerance instanceof Double) {
+                                range = Range.open(((Number) expectedRowField).doubleValue() - (Double) tolerance, ((Number) expectedRowField).doubleValue() + (Double) tolerance);
+                            }
+                            else {
+                                fail("tolerance can only be of Integer, Long, Float or Double Types but found " + tolerance.getClass());
+                            }
```

Comment:
with `.withNumberTolerance` approach, we can collapse this to single `double` case

Reply:
Done.